date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/RecurringTimer.kt

```diff
@@ -0,0 +1,190 @@
+                fun start(
+                    interval: Duration,
+                    coroutineScope: CoroutineScope,
+                    finishCallback: suspend () -> Unit
+                ): Running = Running(
+                    elapsedSoFar = elapsed,
+                    totalDuration = totalDuration,
+                    interval = interval,
+                    coroutineScope = coroutineScope,
+                    finishCallback = finishCallback
+                )
+                internal fun finish(): Finished = Finished(totalDuration)
+            }
+
+            /** Timer is finished. */
+            class Finished(
+                override val totalDuration: Duration
+            ) : NotRunning(totalDuration), Timer.State.NotRunning.Finished
+        }
+
+        /** Timer is running. */
+        class Running(
+            elapsedSoFar: Duration,
+            override val totalDuration: Duration,
+            interval: Duration,
+            private val coroutineScope: CoroutineScope,
+            private val finishCallback: suspend () -> Unit
+        ) : State(), Timer.State.Running, HandleAfterNewStateIsSet<State>, HandleBeforeOldStateIsRemoved<State> {
+            override val elapsed: Flow<Duration> = tickProvider(
+                offset = elapsedSoFar,
+                max = totalDuration,
+                interval = interval
+            )
+            private lateinit var finishJob: Job
+
+            internal suspend fun pause(): NotRunning.Paused {
+                return NotRunning.Paused(elapsed.first(), totalDuration)
+            }
+            internal fun finish(): NotRunning.Finished = NotRunning.Finished(totalDuration)
+
+            override suspend fun beforeOldStateIsRemoved(oldState: State) {
+                finishJob = coroutineScope.launch {
+                    delay(totalDuration - elapsed.first())
+                    finishCallback()
+                }
+            }
+            override suspend fun afterNewStateIsSet(newState: State) {
+                finishJob.cancel()
+            }
+        }
+    }
+}
+
+private fun tickProvider(
+    offset: Duration,
+    max: Duration,
+    interval: Duration
+): Flow<Duration> {
+    val mark = TimeSource.Monotonic.markNow()
+
+    fun newElapsed() = offset + mark.elapsedNow()
+
+    return flow {
+        var expectedElapsed = offset
+        while (true) {
+            // obtain and deliver a new elapsed time value
+            val elapsed = newElapsed()
+            emit(elapsed.coerceAtMost(max))
+            // quit if the timer would reached the max on the next iteration
+            if (max - interval <= elapsed) break
+            // adaptive correction reduces the interval by time spent delivering emitted value
+            val correction = expectedElapsed - newElapsed()
+            // wait for the next iteration
+            delay(interval + correction)
+            expectedElapsed += interval
```

Comment:
There is some interesting behaviour here if `interval = 1000` and the correction is more than that (main thread being blocked for such times is expected behaviour), I'm not even sure what it will do to be honest.

Reply:
you can test this with sleep on the main dispatcher, but if you make val mark = TimeSource.Monotonic.markNow() a default argument (through the chain) instead of a hard coding it in you can throw in a Fake(?) TimeSource.

Reply:
a negative delay would be equal to a zero delay (actually it'd be better to make it explicit). so the timer won't sleep at all hoping the consumer would catch up eventually.

good idea on substituting the real time source. will try it out